apps/web/src/components/costs/thriftiest-costs-calculation.tsx

```diff
@@ -0,0 +1,69 @@
+interface EnergyData {
+    timestamp: string;
+    value: number;
+}
+
+interface UserData {
+    basePrice: number;
+    workingPrice: number;
+}
+
+const formatDate = (date: Date): string => date.toISOString().split("T")[0];
+
+const calculateDailyCosts = (energyData: EnergyData[], userData: UserData[]): Record<string, number> => {
+    if (
+        !energyData ||
+        energyData.length === 0 ||
+        !userData ||
+        userData.length < 2 ||
+        typeof userData[1].basePrice !== "number" ||
+        typeof userData[1].workingPrice !== "number"
+    ) {
+        return {};
+    }
+
+    const dailyCosts: Record<string, number> = {};
+
+    for (const { timestamp, value } of energyData) {
```

Comment:
Du hast diese Gruppierungslogik jetzt sehr oft drin (auch schon in apps/web/src/components/costs/average-costs-calculation.tsx). Kann man da vllt. eine allgemeine Methode ableiten?

Reply:
Jaa zumindest bei den dailyCosts geht das, da hast du Recht

Reply:
Meinte eigentlich eher eine Methode, wo man reingeben kann, wonach man gruppieren (sozusagen die Definition eines Keys). Dann kann man durch die Schleife gehen und für jeden Eintrag den ermitteln und dass dann zum Record hinzufügen. Stelle ich mir so in der Art vor:
```typescript
function groupBy(energyData: EnergyData[], keySelector: (energyData: EnergyData) => string) {
    const grouped: Record<string, number> = {};

    for (const { timestamp, value } of energyData) {
        const date = new Date(timestamp);
        const day = formatDate(date);

        const cost = value * userData[1].workingPrice;
        if (!grouped[day]) {
            grouped[day] = 0;
        }

        grouped[day] += cost;
    }

    return grouped;
}
```
Aber ka ob das funktioniert :smile: